File: Day 1 - Arrays I/73. Set Matrix Zeroes/Solution.py
class Solution:
    def setZeroes(self, matrix: [[]]) -> [[]]:
        row, col = len(matrix), len(matrix[0])
        for i in range(row):
            for j in range(col):
                if matrix[i][j] == 0:
                    self.markCoords(i, j, matrix)

        for i in range(row):
            for j in range(col):
                if matrix[i][j] == -1:
                    matrix[i][j] = 0

        return matrix

    def markCoords(self, i, j, matrix):
        row, col = len(matrix), len(matrix[0])
        for k in range(row):
            if matrix[k][j] != 0:
                matrix[k][j] = -1
        for k in range(col):
            if matrix[i][k] != 0:
                matrix[i][k] = -1
        return matrix


if __name__ == "__main__":
    solution: Solution = Solution()
    print(solution.setZeroes([[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]]))
    # print(solution.markCoords(1, 1, [[1, 1, 1], [1, 0, 1], [1, 1, 1]]))

File: Day 1 - Arrays I/73. Set Matrix Zeroes/Solution.cpp
//
// Created by SHANTANU on 31-05-2023.
//
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    void setZeroes(vector<vector<int>> &matrix) {
        vector<pair<int, int>> zeroes;
        for (int i = 0; i < matrix.size(); i++) {
            for (int j = 0; j < matrix[0].size(); j++) {
                if (matrix[i][j] == 0) {
                    zeroes.emplace_back(i, j);
                }
            }
        }

        for (pair<int, int> item: zeroes) {
            for (int k = 0; k < matrix[0].size(); k++) {
                matrix[item.first][k] = 0;
            }
            for (int k = 0; k < matrix.size(); k++) {
                matrix[k][item.second] = 0;
            }
        }

        for (int i = 0; i < matrix.size(); i++) {
            for (int j = 0; j < matrix[0].size(); j++) {
                printf("%d ", matrix[i][j]);
            }
            cout << endl;
        }
    }
};

int main() {

    vector<vector<int>> matrix = {{0, 1, 2, 0},
                                  {3, 4, 5, 2},
                                  {1, 3, 1, 5}};

    Solution solution;
    solution.setZeroes(matrix);


    return 0;
}